File: packages/core/src/cli/update.ts
import { execFile } from 'node:child_process';
import { existsSync } from 'node:fs';
import { mkdtemp, rm, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { dirname, join } from 'node:path';
import { promisify } from 'node:util';
import { DAEMON_VERSION } from '../version.js';

const execFileAsync = promisify(execFile);

const REPO = 'qlan-ro/mainframe';

export interface GhAsset {
  name: string;
  browser_download_url: string;
}
export interface GhRelease {
  tag_name: string;
  prerelease: boolean;
  draft?: boolean;
  assets: GhAsset[];
}

export interface UpdateOptions {
  /** Specific tag to install (e.g. `v2.0.0-rc.1`). Overrides the latest lookup. */
  version?: string;
  /** Consider pre-releases when picking the newest release. */
  includePrerelease?: boolean;
  /** Install even if the target release is not newer than the running version. */
  force?: boolean;
}

/** Map the running platform to the release artifact filename it should install. */
export function standaloneArtifactName(platform: NodeJS.Platform, arch: string): string {
  const os = platform === 'darwin' ? 'darwin' : platform === 'linux' ? 'linux' : null;
  const cpu = arch === 'x64' ? 'x64' : arch === 'arm64' ? 'arm64' : null;
  if (!os || !cpu) {
    throw new Error(`Unsupported platform for self-update: ${platform}-${arch}`);
  }
  return `mainframe-daemon-${os}-${cpu}.tar.gz`;
}

/** Pick the release to install from a GitHub `/releases` list (newest first). */
export function pickRelease(releases: GhRelease[], opts: UpdateOptions): GhRelease {
  const published = releases.filter((r) => !r.draft);

  if (opts.version) {
    const norm = (t: string) => t.replace(/^v/, '');
    const found = published.find((r) => r.tag_name === opts.version || norm(r.tag_name) === norm(opts.version!));
    if (!found) throw new Error(`No release found for version ${opts.version}`);
    return found;
  }

  const candidates = published.filter((r) => opts.includePrerelease || !r.prerelease);
  const newest = candidates[0]; // GitHub returns releases newest-first
  if (!newest) {
    const hint = published.some((r) => r.prerelease)
      ? ' Only pre-releases exist — retry with `mainframe update --pre`.'
      : '';
    throw new Error(`No matching release found.${hint}`);
  }
  return newest;
}

interface SemVer {
  major: number;
  minor: number;
  patch: number;
  prerelease: string[];
}

function parseSemver(version: string): SemVer {
  const [core, ...preParts] = version.replace(/^v/, '').split('-');
  const [major = 0, minor = 0, patch = 0] = core.split('.').map((n) => Number(n) || 0);
  const prerelease = preParts.join('-');
  return { major, minor, patch, prerelease: prerelease ? prerelease.split('.') : [] };
}

/** Compare two (optionally `v`-prefixed) semver strings: negative if `a` < `b`, 0 if equal, positive if `a` > `b`. */
export function compareSemver(a: string, b: string): number {
  const va = parseSemver(a);
  const vb = parseSemver(b);
  if (va.major !== vb.major) return va.major - vb.major;
  if (va.minor !== vb.minor) return va.minor - vb.minor;
  if (va.patch !== vb.patch) return va.patch - vb.patch;
  // A release with no prerelease suffix outranks one with the same major.minor.patch that has one.
  if (va.prerelease.length === 0 && vb.prerelease.length > 0) return 1;
  if (va.prerelease.length > 0 && vb.prerelease.length === 0) return -1;
  const len = Math.max(va.prerelease.length, vb.prerelease.length);
  for (let i = 0; i < len; i++) {
    const pa = va.prerelease[i];
    const pb = vb.prerelease[i];
    if (pa === undefined) return -1;
    if (pb === undefined) return 1;
    const na = Number(pa);
    const nb = Number(pb);
    const aIsNum = pa !== '' && !Number.isNaN(na);
    const bIsNum = pb !== '' && !Number.isNaN(nb);
    if (aIsNum && bIsNum) {
      if (na !== nb) return na - nb;
    } else if (aIsNum !== bIsNum) {
      return aIsNum ? -1 : 1; // numeric identifiers rank below alphanumeric ones (semver 11.4.4)
    } else if (pa !== pb) {
      return pa < pb ? -1 : 1;
    }
  }
  return 0;
}

/** Refuse to install a release that isn't newer than `currentVersion`, unless `force` is set. */
export function assertNotDowngrade(release: GhRelease, currentVersion: string, force: boolean): void {
  if (force) return;
  if (compareSemver(release.tag_name, currentVersion) > 0) return;
  throw new Error(
    `${release.tag_name} is not newer than the running version (v${currentVersion}). ` +
      'Refusing to downgrade — pass --force to override.',
  );
}

/** Resolve the download URL of `artifact` within a release, or throw. */
export function assetUrl(release: GhRelease, artifact: string): string {
  const asset = release.assets.find((a) => a.name === artifact);
  if (!asset) {
    throw new Error(`Release ${release.tag_name} does not include ${artifact}.`);
  }
  return asset.browser_download_url;
}

/** Parse `update` argv (everything after the subcommand) into options. */
export function parseUpdateArgs(argv: string[]): UpdateOptions & { help?: boolean; dir?: string } {
  const opts: UpdateOptions & { help?: boolean; dir?: string } = {};
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    if (arg === '--pre' || arg === '--prerelease') opts.includePrerelease = true;
    else if (arg === '--force') opts.force = true;
    else if (arg === '-h' || arg === '--help') opts.help = true;
    else if (arg === '--version') opts.version = argv[++i];
    else if (arg === '--dir') opts.dir = argv[++i];
    else throw new Error(`Unknown argument: ${arg}`);
  }
  return opts;
}

/**
 * Locate the standalone install root (the dir holding `bin/` and `lib/`). The
 * wrapper exports MAINFRAME_STANDALONE_ROOT; fall back to deriving it from the
 * bundled node's path (`<root>/bin/node`).
 */
export function resolveInstallRoot(env = process.env, execPath = process.execPath): string {
  const fromEnv = env['MAINFRAME_STANDALONE_ROOT'];
  if (fromEnv && existsSync(join(fromEnv, 'lib', 'daemon.cjs'))) return fromEnv;

  const derived = dirname(dirname(execPath)); // <root>/bin/node → <root>
  if (existsSync(join(derived, 'lib', 'daemon.cjs'))) return derived;

  throw new Error(
    'Could not locate a standalone install to update. This looks like a dev or ' +
      'non-standalone build. Re-run the install script instead:\n' +
      '  curl -fsSL https://raw.githubusercontent.com/qlan-ro/mainframe/main/scripts/install.sh | bash',
  );
}

async function fetchJson<T>(url: string): Promise<T> {
  const headers: Record<string, string> = { 'User-Agent': 'mainframe-updater' };
  if (process.env['GITHUB_TOKEN']) headers['Authorization'] = `Bearer ${process.env['GITHUB_TOKEN']}`;
  const res = await fetch(url, { headers });
  if (!res.ok) throw new Error(`GitHub API ${res.status} for ${url}`);
  return (await res.json()) as T;
}

async function downloadTarball(url: string, dest: string): Promise<void> {
  const res = await fetch(url, { headers: { 'User-Agent': 'mainframe-updater' }, redirect: 'follow' });
  if (!res.ok) throw new Error(`Download failed (${res.status}) for ${url}`);
  await writeFile(dest, Buffer.from(await res.arrayBuffer()));
}

export async function runUpdate(argv: string[] = process.argv.slice(3)): Promise<void> {
  let opts: UpdateOptions & { help?: boolean; dir?: string };
  try {
    opts = parseUpdateArgs(argv);
  } catch (err) {
    console.error(`  ${(err as Error).message}`);
    printUsage();
    process.exit(1);
  }
  if (opts.help) return printUsage();

  const root = opts.dir ?? resolveInstallRoot();
  const artifact = standaloneArtifactName(process.platform, process.arch);

  console.log('  Checking for updates…');
  const listUrl = opts.version
    ? `https://api.github.com/repos/${REPO}/releases/tags/${opts.version}`
    : `https://api.github.com/repos/${REPO}/releases?per_page=30`;
  const payload = await fetchJson<GhRelease | GhRelease[]>(listUrl);
  const release = pickRelease(Array.isArray(payload) ? payload : [payload], opts);
  assertNotDowngrade(release, DAEMON_VERSION, opts.force ?? false);
  const url = assetUrl(release, artifact);

  console.log(`  Installing ${release.tag_name} (${artifact}) → ${root}`);
  const tmp = await mkdtemp(join(tmpdir(), 'mainframe-update-'));
  try {
    const tarball = join(tmp, artifact);
    await downloadTarball(url, tarball);
    // Overwrite bin/ and lib/ in place; the running daemon keeps its loaded copy.
    await execFileAsync('tar', ['-xzf', tarball, '-C', root, '--strip-components=1']);
  } finally {
    await rm(tmp, { recursive: true, force: true });
  }

  console.log(`\n  Updated to ${release.tag_name}.`);
  console.log('  Restart the daemon to run it: `systemctl restart mainframe` (or restart the process).\n');
}

function printUsage(): void {
  console.log(
    [
      '',
      '  mainframe update — upgrade the standalone daemon in place',
      '',
      '  Usage: mainframe update [--pre] [--version <tag>] [--dir <path>] [--force]',
      '',
      '    --pre              include pre-releases when picking the newest',
      '    --version <tag>    install a specific release tag (e.g. v2.0.0-rc.1)',
      '    --dir <path>       install root to unpack over (default: auto-detected)',
      '    --force            allow installing a release that is not newer than the current one',
      '',
    ].join('\n'),
  );
}

File: packages/core-rs/crates/mainframe-daemon/src/cli/update.rs
//! Ported from `packages/core/src/cli/update.ts` — `mainframe update`.
//!
//! Self-updates a standalone install in place: picks a GitHub release, refuses
//! a downgrade unless `--force`, downloads the matching platform tarball, and
//! extracts it over the install root with a `tar` shell-out (matching the TS
//! twin's `execFile('tar', ...)`). Release-picking/semver logic lives in
//! `release` (pure, unit-tested); this module owns argv parsing, install-root
//! resolution, and the network/process orchestration.

mod release;

use std::collections::HashMap;
use std::path::{Path, PathBuf};

use release::{assert_not_downgrade, asset_url, pick_release, standalone_artifact_name};
use serde::Deserialize;

const REPO: &str = "qlan-ro/mainframe";

/// Parsed `mainframe update` flags (`packages/core/src/cli/update.ts`'s `UpdateOptions`).
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct UpdateOptions {
    pub version: Option<String>,
    pub include_prerelease: bool,
    pub force: bool,
    pub help: bool,
    pub dir: Option<String>,
}

/// `parseUpdateArgs()`.
pub fn parse_update_args(argv: &[String]) -> Result<UpdateOptions, String> {
    let mut opts = UpdateOptions::default();
    let mut i = 0;
    while i < argv.len() {
        match argv[i].as_str() {
            "--pre" | "--prerelease" => opts.include_prerelease = true,
            "--force" => opts.force = true,
            "-h" | "--help" => opts.help = true,
            "--version" => {
                i += 1;
                opts.version = argv.get(i).cloned();
            }
            "--dir" => {
                i += 1;
                opts.dir = argv.get(i).cloned();
            }
            other => return Err(format!("Unknown argument: {other}")),
        }
        i += 1;
    }
    Ok(opts)
}

/// `resolveInstallRoot()`. The standalone layout has no `lib/` anymore (T5.3
/// drops the Node bundle), so the marker is the daemon binary itself:
/// `bin/mainframe-daemon` beside the `mainframe` wrapper.
pub fn resolve_install_root(
    env: &HashMap<String, String>,
    exe_path: &Path,
) -> Result<PathBuf, String> {
    if let Some(root) = env.get("MAINFRAME_STANDALONE_ROOT") {
        let candidate = PathBuf::from(root);
        if is_standalone_root(&candidate) {
            return Ok(candidate);
        }
    }

    if let Some(derived) = exe_path.parent().and_then(Path::parent)
        && is_standalone_root(derived)
    {
        return Ok(derived.to_path_buf());
    }

    Err(
        "Could not locate a standalone install to update. This looks like a dev or \
         non-standalone build. Re-run the install script instead:\n  curl -fsSL \
         https://raw.githubusercontent.com/qlan-ro/mainframe/main/scripts/install.sh | bash"
            .to_string(),
    )
}

fn is_standalone_root(root: &Path) -> bool {
    root.join("bin").join("mainframe-daemon").is_file()
}

fn platform_name() -> &'static str {
    match std::env::consts::OS {
        "macos" => "darwin",
        other => other,
    }
}

fn arch_name() -> &'static str {
    match std::env::consts::ARCH {
        "aarch64" => "arm64",
        "x86_64" => "x64",
        other => other,
    }
}

/// `runUpdate()`.
pub async fn run_update(argv: Vec<String>) -> Result<(), String> {
    let opts = parse_update_args(&argv).inspect_err(|_| print_usage())?;
    if opts.help {
        print_usage();
        return Ok(());
    }

    let root = resolve_root(&opts)?;
    let artifact = standalone_artifact_name(platform_name(), arch_name())?;

    println!("  Checking for updates…");
    let releases = fetch_releases(opts.version.as_deref()).await?;
    let release = pick_release(&releases, opts.version.as_deref(), opts.include_prerelease)?;
    assert_not_downgrade(&release, crate::DAEMON_VERSION, opts.force)?;
    let url = asset_url(&release, &artifact)?;

    println!(
        "  Installing {} ({artifact}) -> {}",
        release.tag_name,
        root.display()
    );
    install_release(&url, &artifact, &root).await?;

    println!("\n  Updated to {}.", release.tag_name);
    println!("  Restart the daemon for the update to take effect.\n");
    Ok(())
}

fn resolve_root(opts: &UpdateOptions) -> Result<PathBuf, String> {
    if let Some(dir) = &opts.dir {
        return Ok(PathBuf::from(dir));
    }
    let env: HashMap<String, String> = std::env::vars().collect();
    let exe = std::env::current_exe()
        .map_err(|err| format!("Could not resolve the running binary path: {err}"))?;
    resolve_install_root(&env, &exe)
}

// GitHub returns an array from `/releases` but a single object from
// `/releases/tags/<tag>` — untagged covers both response shapes.
#[derive(Deserialize)]
#[serde(untagged)]
enum ReleasesPayload {
    List(Vec<release::GhRelease>),
    Single(Box<release::GhRelease>),
}

async fn fetch_releases(version: Option<&str>) -> Result<Vec<release::GhRelease>, String> {
    let url = match version {
        Some(v) => format!("https://api.github.com/repos/{REPO}/releases/tags/{v}"),
        None => format!("https://api.github.com/repos/{REPO}/releases?per_page=30"),
    };
    let mut req = reqwest::Client::new()
        .get(&url)
        .header("User-Agent", "mainframe-updater");
    if let Ok(token) = std::env::var("GITHUB_TOKEN") {
        req = req.header("Authorization", format!("Bearer {token}"));
    }
    let res = req
        .send()
        .await
        .map_err(|err| format!("GitHub API request failed for {url}: {err}"))?;
    if !res.status().is_success() {
        return Err(format!("GitHub API returned {} for {url}", res.status()));
    }
    let payload: ReleasesPayload = res
        .json()
        .await
        .map_err(|err| format!("Failed to parse the GitHub API response: {err}"))?;
    Ok(match payload {
        ReleasesPayload::List(list) => list,
        ReleasesPayload::Single(release) => vec![*release],
    })
}

async fn install_release(url: &str, artifact: &str, root: &Path) -> Result<(), String> {
    let tmp_dir = std::env::temp_dir().join(format!("mainframe-update-{}", nanoid::nanoid!(8)));
    tokio::fs::create_dir_all(&tmp_dir)
        .await
        .map_err(|err| format!("Could not create a temp dir for the download: {err}"))?;
    let tarball = tmp_dir.join(artifact);

    let result = match download_tarball(url, &tarball).await {
        Ok(()) => extract_over(&tarball, root).await,
        Err(err) => Err(err),
    };

    let _ = tokio::fs::remove_dir_all(&tmp_dir).await; // best-effort cleanup /* expected */
    result
}

async fn download_tarball(url: &str, dest: &Path) -> Result<(), String> {
    let res = reqwest::Client::new()
        .get(url)
        .header("User-Agent", "mainframe-updater")
        .send()
        .await
        .map_err(|err| format!("Download failed for {url}: {err}"))?;
    if !res.status().is_success() {
        return Err(format!("Download failed ({}) for {url}", res.status()));
    }
    let bytes = res
        .bytes()
        .await
        .map_err(|err| format!("Download failed while reading {url}: {err}"))?;
    tokio::fs::write(dest, &bytes)
        .await
        .map_err(|err| format!("Could not write {}: {err}", dest.display()))?;
    Ok(())
}

// Extraction shells out to the system `tar` (matches the TS twin's
// `execFile('tar', ...)`) rather than pulling in a Rust tar/gzip crate.
async fn extract_over(tarball: &Path, root: &Path) -> Result<(), String> {
    let status = tokio::process::Command::new("tar")
        .arg("-xzf")
        .arg(tarball)
        .arg("-C")
        .arg(root)
        .arg("--strip-components=1")
        .status()
        .await
        .map_err(|err| format!("Failed to run tar: {err}"))?;
    if !status.success() {
        return Err(format!("tar exited with status {status}"));
    }
    Ok(())
}

fn print_usage() {
    println!(
        "\n  mainframe update — upgrade the standalone daemon in place\n\n  Usage: mainframe update [--pre] [--version <tag>] [--dir <path>] [--force]\n\n    --pre              include pre-releases when picking the newest\n    --version <tag>    install a specific release tag (e.g. v2.0.0-rc.1)\n    --dir <path>       install root to unpack over (default: auto-detected)\n    --force            allow installing a release that is not newer than the current one\n"
    );
}

#[cfg(test)]
mod tests;
